### ISSUE TITLE
Feature Request - More Details

### ISSUE THREAD

**Issue by brycesub** [justinleewells/pogo-optimizer#43]
A few features I think would be helpful:
Show the level of the Pokemon.
Show the resource cost of powering up 100% (at the player's current level, or a configurable player level).
Show the most desired move-set for the final evolution of the Pokemon (maybe too subjective?)

Thoughts?

**Comment by cstrachan88** [comment]
I agree. I think a combo of this, and [this](http://squarecat.io/pokemongo-data/#/pokedex?_k=lw1qbm) would be perfect.

**Comment by jnsnow** [comment]
@brycesub How do you compute the "level"? It's not part of the raw data you get from the Inventory event. (Only CP, the IVs, and how many powerups you've used.)

**Comment by justinleewells** [comment]
Level may be able to be included in the future, but for now it isn't feasible. Same with powering up to 100%. I had intended to eventually give suggestions on movesets, however. Additionally, I'd like to include the ability to "sort by DPS" and "sort by durability" in the future.

**Comment by eratio08** [comment]
you can compute the range of level by the dust-costs (is this number served by the data?)

> level	Stardust
> 1	200
> 2	200
> 3	200
> 4	200
> 5	400
> 6	400
> 7	400
> 8	400
> 9	600
> 10	600
> 11	600
> 12	600
> 13	800
> 14	800
> 15	800
> 16	800
> 17	1000
> 18	1000
> 19	1000
> 20	1000
> 21	1300
> 22	1300
> 23	1300
> 24	1300
> 25	1600
> 26	1600
> 27	1600
> 28	1600
> 29	1900
> 30	1900
> 31	1900
> 32	1900
> 33	2200
> 34	2200
> 35	2200
> 36	2200
> 37	2500
> 38	2500
> 39	2500
> 40	2500
> 41	3000
> 42	3000
> 43	3000
> 44	3000
> 45	3500
> 46	3500
> 47	3500
> 48	3500
> 49	4000
> 50	4000
> 51	4000
> 52	4000
> 53	4500
> 54	4500
> 55	4500
> 56	4500
> 57	5000
> 58	5000
> 59	5000
> 60	5000
> 61	6000
> 62	6000
> 63	6000
> 64	6000
> 65	7000
> 66	7000
> 67	7000
> 68	7000
> 69	8000
> 70	8000
> 71	8000
> 72	8000
> 73	9000
> 74	9000
> 75	9000
> 76	9000
> 77	10000
> 78	10000
> 79	10000
> 80	10000

https://docs.google.com/spreadsheets/d/1q_GTEoPUpoES54XNDFRENfyd1TZsSkqugxkO_4IbV3Y/edit

**Comment by justinleewells** [comment]
I'd honestly rather wait until I can add in a hook to check power ups and do it the right way the first time. I'll be re-writing the inventory system in the near future and that should be pretty easy to do at the same time. I've just been sidetracked with this compatibility stuff lately.

**Comment by basmith7** [comment]
Level can be calculated. 

The CP formula is given here http://pokemongo.gamepress.gg/pokemon-stats-advanced#calculating-cp

Cp multiplier is level dependant. http://pokemongo.gamepress.gg/cp-multiplier

**Comment by justinleewells** [comment]
Very interesting information. I'll look into getting levels added this weekend.

**Comment by jnsnow** [comment]
Dust cost isn't served by the GetInventory callback afaict but it does show you how many times you've powered up the Pokemon itself.

**Comment by justinleewells** [comment]
Very true. I wonder how the game gets the information on level up cost though.

**Comment by jnsnow** [comment]
Alright, I'm looking at my Vaporeon here with 15/13/14 (ATK/DEF/STA), CP918.

CP = (ATK * DEF^0.5 * STA^0.5 * CPM^2) / 10
918 = ((VATK + 15) * (VDEF + 13)^0.5 * (VSTA + 14)^0.5 * CPM^2) / 10

Looking at the Master Protobuf rip that's floating around:

    Stats {
      BaseStamina: 260
      BaseAttack: 186
      BaseDefense: 168
    }

918 = ((186+15) * (168+13)^0.5 + (260+14)^0.5 * CPM^2) / 10
918 = (2720.73 * CPM^2) / 10
9180 = (2720.73 * CPM^2)
3.374 = CPM^2
CPM = sqrt(3.374) = 1.83687

Uh, well, that seems wrong.
Any feedback on where I may have went wrong?

(Note: I *did* upgrade this Vaporeon 3 times, it is now at a cost of 1300.
Based on the dust cost chart, that'd be "21-24". I believe other charts use "half levels," so this would be level 10.5 to 12. Based on the CPM chart, that should be .43 to .46.)

= (186+15 * (168+13)^0.5 + (260+14)^0.5 * .445) / 10
= (186+15 * (168+13)^0.5 + (260+14)^0.5 * .445) / 10
= 39.51.

Err, yeah, this doesn't seem right.
Tips?

**Comment by basmith7** [comment]
Attack = (BaseAttack+ IndividualAttack) * CPM
Defense = (BD + ID) * CPM
Stamina = (BS + IS) * CPM

CP = ((Attack) * (Defense)^0.5 * (Stamina)^0.5 * (CPM)^2) / 10

CP = (  ((BA+IA)*CPM)  *  ((BD+ID)*CPM)^0.5  *  ((BS+IS)*CPM)^0.5  *  (CPM)^2) / 10

https://www.wolframalpha.com/input/?i=918+%3D+(((186+%2B+15)+*+x)+*+((168+%2B+13)+*+x)%5E0.5+*+((260+%2B+14+)+*+x)%5E0.5+*x%5E2)+%2F+10

**Comment by justinleewells** [comment]
Any idea if you can just round up/down or how that would work when it lies between two values?

**Comment by basmith7** [comment]
Something is off because that's still not matching the stardust cost...

**Comment by jnsnow** [comment]
Oh, each individual term is multiplied by the CPM? That's not really clear from the way that page writes it, but OK. It's also *definitely different* from the formula they wrote here: 

> "Max  CP = ((Base_Attack + 15) * ((Base_Defense + 15)^0.5 * (Base_Stamina + 15)^0.5 * 0.7903001^2) / 10"

But okay, let me re-run my numbers.

ATK = (186 + 15) * X
DEF = (168 + 13) * X
STA = (260 + 14) * X

CP = [((BA + IVA) * CPM) * ((BD + IVD) * CPM)^0.5 * ((BS + IVS) * CPM)^0.5 * X^2] / 10
CP = [((186 + 15) * X) * ((168 + 13) * X)^0.5 * ((260 + 14) * X)^0.5 * X^2] / 10
918 = [(201 * X) * (181 * X)^0.5 * (274 * X)^0.5 * X^2] / 10
9180 = (201 * X) * (181 * X)^0.5 * (274 * X)^0.5 * X^2
9180 = 201 * sqrt(49594) * x^4
9180 / (201 * sqrt(49594)) = x^4
~0.205084 = x^4
x ~= 0.67295

This would also imply the real formula to be something like:

CP = [(BASE_ATK + ATK_IV) * sqrt((BASE_DEF + DEF_IV) * (BASE_STA + STA_IV)) * CPM^4] / 10
or
CPM^4 = (CP * 10) / [(BASE_ATK + ATK_IV) * sqrt((BASE_DEF + DEF_IV) * (BASE_STA + STA_IV))]
CPM = {(CP * 10) / [(BASE_ATK + ATK_IV) * sqrt((BASE_DEF + DEF_IV) * (BASE_STA + STA_IV))]}^0.25

which, just to confirm I am not /the worst/ at math:

= (918 * 10) / [(186 + 15) * sqrt((168 + 13) * (260 + 14))]
= 9180 / [(201 * sqrt((181) * (274))]
CPM^4 = 9180 / [201 * sqrt(49594)]
~= 0.205084
CPM ~= (0.205084)^0.25 ~= 0.67295

So my reduction is OK.

Again, powerup cost is at 1300 so by these charts that's level "11-13." which is only a CPM of 0.44 through 0.48, so we're way off course here.

Alright, at least that's a sane multiplier... but it's still pretty wrong... Even if they rounded those numbers, it shouldn't be off by ***that much***. I am not inclined to really trust this formula...

*Feeds Vaporeon a candy*

+40, CP now 958. Cost still 1300.
+40, CP now 998. Cost still 1300!
+40, CP now 1038. Cost is now 1600.

If I'm reading the dust cost chart right, I'm now at Step 25 that displays 1600, so I must have been at Step 22.

Let's re-run the numbers...

CPM = {(CP * 10) / [(BASE_ATK + ATK_IV) * sqrt((BASE_DEF + DEF_IV) * (BASE_STA + DEF_IV))]}^0.25
CPM = 0.693940

Chart puts that at "Level 27," which is a stardust cost of 4000 or so... I don't think these formula are correct.

**Comment by jnsnow** [comment]
Looks like the parentheses on the source provided was incorrect.

Real formula is apparently:

CP = (BA + IVA) * (BD + IVD)^0.5 * (BS + IVS)^0.5 * (CPM^2 / 10)

Difference is in what's getting divided by 10. It's just the CPM^2, not the whole enchilada.

**deep breath**

1038 = (186 + 15) * (168 + 13)^0.5 * (260 + 14)^0.5 * (CPM^2/10)
1038 = 201 * sqrt(49594) * CPM^2/10
10380 / [201 * sqrt(49594)] = CPM^2
CPM = sqrt(10380 / [201 * sqrt(49594)])
CPM ~= 0.48155219 ...

Power-up cost: 1600 -- 25th power-up state.
Level as given by CPM -- nearest match -- 0.48168495 is Level 13.

Whoa! Direct hit!

(The 25th power-up state, what I'd really like to call level because I think half-levels are unnecessarily complicated, is the same as Level 13.)

:)

**Comment by justinleewells** [comment]
So we can actually use this to determine levels. Great success! Thanks for all your work.

**Comment by jnsnow** [comment]
Yes, though unfortunately the last step there is an approximate table lookup instead of a direct conversion, but with FLOPS errors that's probably as close as you're going to get.

Final formula looks something like this:

CPM = [(CP * 10) / {(ATK_B + ATK_IV) * [(DEF_B + DEF_IV) * (STA_B + STA_IV)]^0.5}]^0.5

e.g., my Vaporeon again (1038CP):

CPM = [(1038 * 10) / {(186 + 15) * [(168 + 13) * (260 + 14)]^0.5}]^0.5
CPM ~= 0.481552191487 ...

Likely the rounding is due to the CP being rounded up/down from its actual value, which is likely a float of some sort.

Also found these formula which may be of interest:
https://drive.google.com/file/d/0B0TeYGBPiuzaenhUNE5UWnRCVlU/view

Includes a lot of useful stuff for us,
HP, ATK, DEF,
the DPS which another issue requested.

I also like the idea of having:

%perfect (As it is calculated now, a straight percentage)
%cps (Weighted IV perfection based on maximizing CPS)

Lastly, all known CPM can fit on a 5-degree polynomial:

CPM = 1.278E-9x^5 - 2.981E-7x^4 + 2.598E-5x^3 - 1.112E-3x^2 + 0.032x + 0.075

(for X as the Nth power up state, i.e. where my Vaporeon is Lv25.)

Taking our computed CPM, we can solve for X (Well, wolfram can; I am not really adept at fifth degree equations) --

0.48155219 = 1.278E-9x^5 - 2.981E-7x^4 + 2.598E-5x^3 - 1.112E-3x^2 + 0.032x + 0.075
x = 24.9532

So we have calculated that this Vaporeon is power up level 25 from principal stats and a fifth-degree polynomial equation determined via regression from all known data-mined CPM levels for Pokemon power up states 1-80 (i.e. "level 1-40")

My math is literally nowhere good enough to boil that down into y = f(x), though. Lookup table is the best we can do until someone smarter than me can compute level = f(CPM).

This weekend's gonna be a fun one!

--js

**Comment by justinleewells** [comment]
This is excellent work, man. I appreciate your effort.

This weekend will be fun, indeed. I'm going to completely redesign the UI and this data gives us the ability to implement so many new features.

**Comment by basmith7** [comment]
I put together a spreadsheet to figure some things out. Just fill out the info in the second row.

https://docs.google.com/spreadsheets/d/1thedZjuGNQHLY8hmoijURLZrwnBFjoHHZS23vFvRfos/edit?usp=sharing

**Comment by jnsnow** [comment]
BTW, I upgraded under the proxy while dumping everything. The UpgradePokemon event actually shows you the CM:

UpgradePokemon::{
  "result": "SUCCESS",
  "upgraded_pokemon": {
    "id": "10677862168141413653",
    "pokemon_id": "VAPOREON",
    "cp": 1078,
    "stamina": 134,
    "stamina_max": 134,
    "move_1": "WATER_GUN_FAST",
    "move_2": "HYDRO_PUMP",
    "height_m": 1.0086523294448853,
    "weight_kg": 3.386974573135376,
    "individual_attack": 15,
    "individual_defense": 13,
    "individual_stamina": 14,
    "cp_multiplier": 0.42250001430511475,
    "pokeball": "ITEM_POKE_BALL",
    "captured_cell_id": "9935993010607095808",
    "creation_time_ms": "1468702345099",
    "num_upgrades": 7,
    "additional_cp_multiplier": 0.06835579872131348,
    "nickname": "Rainer93"
  }
}

This upgrade pushed me to 1078 as you can see,
which by formula is:

= [(1078 * 10) / {(186 + 15) * [(168 + 13) * (260 + 14)]^0.5}]^0.5
~- 0.4907

The upgrade formula shows this as:

    "cp_multiplier": 0.42250001430511475,
    "additional_cp_multiplier": 0.06835579872131348,

which would be:

0.42250001430511475 + 0.06835579872131348
= 0.49085581302642823

which is the *exact* CPM of this Pokemon.

(cp_multiplier is the CPM when you caught it, additional_cp_multiplier is all of the CPM you've added.)

**Comment by justinleewells** [comment]
You're right. It seems to have come with the pogo-mitm update. I've got levels being added to the metadata right now, though, so we're good to go!
